### PR TITLE
Fix partial reward margins in collator

### LINKED ISSUE #5539: fix: DataCollatorForRewardModelingDataset checks margin only on examples[0], causing silent data loss after shuffle
## Problem

In `trl/trainer/reward_trainer.py`, `DataCollatorForRewardModelingDataset.torch_call()` checks for the `"margin"` key only on `examples[0]`:

```python
if "margin" in examples[0]:
    margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)
```

This has two failure modes:

1. **Silent data loss**: After dataset shuffling, if the first example lacks `"margin"` but subsequent examples have it, margins are silently dropped and the margin-aware reward loss is skipped with no warning.
2. **Runtime `KeyError`**: If the first example has `"margin"` but a later one does not (inconsistent dataset), the list comprehension crashes.

## Expected Behavior

All examples should be checked for the `"margin"` key. Missing margins should default to `0.0` gracefully.

## Fix

```python
if any("margin" in example for example in examples):
    margins = torch.tensor([example.get("margin", 0.0) for example in examples], dtype=torch.float)
```

### PR DESCRIPTION
**Pull request by he-yufeng**
﻿# What does this PR do?

Fixes #5539.

`DataCollatorForPreference` currently decides whether to emit a `margin` tensor by checking only `examples[0]`. That makes the result depend on batch order:

- if the first example has no `margin`, later margins are silently dropped
- if the first example has `margin` but a later example does not, collation raises `KeyError`

This PR checks the whole batch instead. If any example provides `margin`, the collator emits a margin tensor and defaults missing margins to `0.0`, matching the optional-field behavior documented for this collator.

I added regression tests for both mixed-order cases.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

## Tests

```text
python -m pytest tests/test_reward_trainer.py::TestDataCollatorForPreference -q
python -m ruff check trl\trainer\reward_trainer.py tests\test_reward_trainer.py
python -m py_compile trl\trainer\reward_trainer.py tests\test_reward_trainer.py
git diff --check
```

---

> [!NOTE]
> **Medium Risk**
> Changes reward-training batch collation and can alter loss when batches mix examples with and without margins; scope is small and covered by new tests.
> 
> **Overview**
> Fixes **`DataCollatorForPreference`** so optional **`margin`** values are handled correctly when only some examples in a batch define them.
> 
> Previously, collation keyed off **`examples[0]`** only: margins on later rows could be dropped, or a missing key on a later row could raise **`KeyError`**. The collator now emits a **`margin`** tensor if **any** example has **`margin`**, and fills missing entries with **`0.0`**.
> 
> Two unit tests cover mixed batches (margin missing on the first vs. on a later example).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5988fbb8ea7d3b5ea71401b988ce1b0b45dbbc2b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
thanks, can you answer this before: https://github.com/huggingface/trl/issues/5539#issuecomment-4255437708

**Comment by he-yufeng** [comment]
Answered on the issue here: https://github.com/huggingface/trl/issues/5539#issuecomment-4618755950

Short version: I don't have a canonical paper example for intentionally mixed-margin rows. The practical issue is that `margin` is optional per example, but the current collator makes the batch decision from only `examples[0]`, so behavior depends on shuffle/order. I also noted that I can switch the PR to a clear `ValueError` for mixed batches if you prefer stricter semantics.
